File: learning-dsa/classical-implementations/ds/bst/BST/binarysearchtree.cpp
#include "binarysearchtree.h"
#include <iostream>

BinarySearchTree::BinarySearchTree()
{
    std::cout << "Calling constructor" << std::endl;
    initializeBinarySearchTree();
}

BinarySearchTree::BinarySearchTree(BinarySearchTree const &)
{
    std::cout << "Calling copy constructor" << std::endl;
}

BinarySearchTree& BinarySearchTree::operator=(BinarySearchTree const &)
{
    std::cout << "Calling operator =" << std::endl;
}

BinarySearchTree::~BinarySearchTree()
{
    std::cout << "Calling destructor" << std::endl;
    clearBinarySearchTree();
}

void BinarySearchTree::initializeBinarySearchTree()
{
    this->root = NULL;
}

void BinarySearchTree::insert(int x)
{
    Node* tmp = this->root;
    Node* tmpParent = NULL;
    bool found = true;
    while(tmp != NULL && !found)
    {
        tmpParent = tmp;
        if(x < tmp->data)
            tmp = tmp->left;
        else if( x > tmp->data)
            tmp = tmp->right;
        else found = true;
    }
    if (fount) return;

    if(tmpParent == NULL)
    {
        Node* node = new Node(x, NULL, NULL, NULL);
        this->root = node;
    }
    else
    {
        Node* node = new Node(x, NULL, NULL, tmpParent);
        if(x < tmpParent->data) tmpParent->left = node;
        else tmpParent->right = node;
    }
}

Node* BinarySearchTree::search(int data)
{
    return NULL;
}

void BinarySearchTree::clearBinarySearchTree()
{
    clearBinarySearchTreeHelper(this->root);
}

void BinarySearchTree::clearBinarySearchTreeHelper(Node* node)
{
    clearBinarySearchTreeHelper(node->left);
    clearBinarySearchTreeHelper(node->left);
    delete node;
}

void BinarySearchTree::inOrderWalk()
{
    inOrderWalkHelper(this->root);
    std::cout << std::endl;
}

void BinarySearchTree::inOrderWalkHelper(Node* node)
{
    if (node == NULL) return;
    inOrderWalkHelper(node->left);
    std::cout << node->data << " ";
    inOrderWalkHelper(node->right);
}

File: learning-python/playground/ds/binary_search_tree.py
class Node:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, x):
        current_node = self.root
        previous_node = None
        while current_node is not None:
            previous_node = current_node
            if x < current_node.data:
                current_node = current_node.left
            else:
                current_node = current_node.right
        if previous_node is None:
            self.root = Node(x)
        elif x < previous_node.data:
            previous_node.left = Node(x)
        else:
            previous_node.right = Node(x)

    def print_tree_left_root_right(node):
        if node is None:
            return
        else:
            print_tree_left_root_right(node.left)
            print(node.data)
            print_tree_left_root_right(node.right)

    def print_tree_left_right_root(node):
        if node is None:
            return
        else:
            print_tree_left_right_root(node.left)
            print_tree_left_right_root(node.right)
            print(node.data)

    def print_tree_root_left_right(node):
        if node is None:
            return
        else:
            print(node.data)
            print_tree_root_left_right(node.left)
            print_tree_root_left_right(node.right)

    def get_minimum(self):
        current_node = self.root
        previous_node = None
        while current_node is not None:
            previous_node = current_node
            current_node = current_node.left
        if previous_node is None:
            print("tree empty")
        else:
            return previous_node.data

    def get_maximum(self):
        current_node = self.root
        previous_node = None
        while current_node is not None:
            previous_node = current_node
            current_node = current_node.right
        if previous_node is None:
            print("tree empty")
        else:
            return previous_node.data

    def search(self, x):
        current_node = self.root
        while current_node is not None and x != current_node.data:
            if x < current_node.data:
                current_node = current_node.left
            else:
                current_node = current_node.right
        if current_node is None:
            print "Element not found"
        else:
            return x

bt = BinarySearchTree()
bt.insert(3)
bt.insert(5)
bt.insert(2)
bt.insert(9)
print(bt.search(3))
bt.print_tree()
print(bt.get_minimum())
print(bt.get_maximum())
bt.insert(3)
bt.insert(
    jj
